File: longest-substring-with-at-most-k-distinct-characters/longest-substring-with-at-most-k-distinct-characters.java
class Solution {
    public int lengthOfLongestSubstringKDistinct(String s, int k) {
        Map<Character, Integer> freq = new HashMap<>();
        int left = 0, max = 0;
        for(int right = 0; right < s.length(); right++){
            freq.put(s.charAt(right), freq.getOrDefault(s.charAt(right), 0) + 1);
            while(freq.size() > k){
                freq.put(s.charAt(left), freq.getOrDefault(s.charAt(left), 1) - 1);
                if(freq.get(s.charAt(left)) == 0) freq.remove(s.charAt(left));
                left++;
            }
            max = Math.max(max, right - left + 1);
        }
        return max;
    }
}

File: longest-substring-with-at-most-k-distinct-characters/longest-substring-with-at-most-k-distinct-characters.go
func lengthOfLongestSubstringKDistinct(s string, k int) int {
    freq := map[byte]int{}
    left,  maxSize := 0, 0
    for right := 0; right < len(s); right++ {
        if _, ok := freq[s[right]]; !ok {
            freq[s[right]] = 1
        }else{
            freq[s[right]]++
        }
        for len(freq) > k {
            if freq[s[left]] == 1 {
                delete(freq, s[left])
            }else{
                freq[s[left]]--
            }
            left++
        }
        maxSize = max(maxSize, right - left + 1)
    }
    return maxSize
}

func max(a, b int) int {
    if a >= b {
        return a
    }
    return b
}